From: "Gordon Farquharson" <gordonfarquharson@gmail.com>

Align the members of the SSB device structure to a 32 bit boundary so that
the b43 driver can be built for arm using a cross compiler.  This alignment
is required so that the test in scripts/mod/file2alias.c that checks that
the size of the device ID type against the size of the section in the
object file succeeds (see comment and http://lkml.org/lkml/2008/2/18/481
for explanation).

Signed-off-by: Gordon Farquharson <gordonfarquharson@gmail.com>
Cc: Russell King <rmk@arm.linux.org.uk>
Acked-by: Michael Buesch <mb@bu3sch.de>
Cc: "John W. Linville" <linville@tuxdriver.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/mod_devicetable.h |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff -puN include/linux/mod_devicetable.h~fix-b43-driver-build-for-arm include/linux/mod_devicetable.h
--- a/include/linux/mod_devicetable.h~fix-b43-driver-build-for-arm
+++ a/include/linux/mod_devicetable.h
@@ -351,7 +351,13 @@ struct sdio_device_id {
 struct ssb_device_id {
 	__u16	vendor;
 	__u16	coreid;
-	__u8	revision;
+	/* Explicit padding to support a broken sanity check in file2alias.c.
+	 * The check compares the size of the structure in the kernel
+	 * object file to the size of the structure reported in userspace for
+	 * the system on which the kernel is compiled. The check breaks on
+	 * cross-compilation, and the padding is a workaround for this. */
+	__u8	revision
+		__attribute__((aligned(sizeof(__u32))));
 };
 #define SSB_DEVICE(_vendor, _coreid, _revision)  \
 	{ .vendor = _vendor, .coreid = _coreid, .revision = _revision, }
_
